notification: support new parameters

dusek opened the pull request:
This PR adds support for these `notification` action parameters: `open`, `execute`, `activate`, `app_icon` and `content_image`. Their meaning is described in the [upstream options documentation](https://github.com/julienXX/terminal-notifier#options).

Parameters prefer underscores (unlike upstream) to be consistent with the fastlane "look and feel".

Params passing refactoring done which should greatly simplify adding new options in the future (used for the above options already).

WARNING: first non-bugfix contribution to fastlane, very newbie in Ruby :smile:

dusek committed:
Support additional parameters in `notification`

Support `open`, `execute`, `activate`, `app_icon`, `content_image`
parameters in the `notification` action.

For the app_icon and content_image parameters, we use underscores in the
parameter names (unlike the upstream terminal-notifier utility), in
order to be consistent with the fastlane "look and feel".

KrauseFx commented on fastlane/docs/Actions.md:
```diff
@@ -1859,6 +1859,13 @@ Display a notification using the OS X notification centre. Uses [terminal-notifi
 
 [ByMyEyes](https://github.com/fastlane/examples/blob/master/BeMyEyes/Fastfile) uses the `notify` action to show a success message after `fastlane` finished executing.
 
+Or perhaps you can use this to open the build details in iTunes Connect by cliking the notification after fastlane finishes the upload (replace `:your_app_id` with App ID of the app on iTunes Connect):
```
`cliking` spelling mistake

KrauseFx replied:
This example is actually what `watchbuild` is for

KrauseFx commented on fastlane/docs/Actions.md:
````diff
@@ -1859,6 +1859,13 @@ Display a notification using the OS X notification centre. Uses [terminal-notifi
+
+```ruby
+  pilot
+  notification(subtitle: "Finished Uploading App", message: "Click to open the build in iTunes Connect", open: "http://itunesconnect.apple.com/WebObjects/iTunesConnect.woa/ra/ng/app/:your_app_id/activity/ios/builds/#{get_version_number}/#{get_build_number}/details")
````
This would need some formatting to multiple lines.

KrauseFx commented:
This looks nice, thanks. I added some comments. Could you also add some tests for this? :+1:

dusek commented:
@KrauseFx sorry for the delay, but I am new to Ruby, and on top of that RSpec is quite a beast :smile: so it took me a little while to find enough time to write tests.

I addressed your review comments:

- *documentation example was analogous to `watchbuild`*: I was not aware of that; in the end I removed the example, I don't think it is good to suggest something in examples that is better handled by a dedicated tool, and I could not think of some other good example to include in the docs
- *typo in docs* - not necessary to fix, as the docs change was removed (see previous point)
- *tests*: I added some tests for the `notification` action (the new commit), mainly based on checking that when `notification` is used in `Fastfile`, then proper arguments are passed to `TerminalNotifier.notify`. Since I am new to Ruby and especially RSpec, I advise caution during their review :smile: 

Writing the tests was actually useful and led to one code change - CircleCI uses Ruby 2.0 (which is what the stock OS X ruby is e.g. in El Capitan), and I was using `Array.to_h` which was introduced only in 2.1 (I was developing on 2.2.4, so I did not notice before). So I worked around it and now this runs on 2.0 too (had I not written the tests, I would not have noticed and if it got released... :smile:)

Please let me know if you have some other review comments, and/or if you want me to squash the commits (btw. GitHub released ["Squash and merge" support for PRs](https://github.com/blog/2141-squash-your-commits) 2 days ago :wink: I tried it and it rebases on top of the latest master and does not create a merge commit, which may or may not be a problem).

KrauseFx commented:
Squashing on GitHub doesn't work for us, since our master branch is protected, so please do squash for us

KrauseFx commented:
Great, thanks for updating the PR, these are some great changes! Have you tried it locally for your project too? :+1:

dusek commented:
@KrauseFx yes, I just tried it with all possible parameters, works nice.

I just pushed all the commits squashed (and rebased on top of the latest master).

KrauseFx commented:
Looks great, but would like a second pair of :eyes: just to be sure :+1: Ready from my side

asfalcone commented:
Thanks @dusek Looks good to me!
